metrics/prometheus/prometheus_test.go

```diff
@@ -0,0 +1,356 @@
+		},
+		{
+			description: "Remove a connection from the open connection count",
+			testCase: func(pm *PrometheusMetrics) {
+				pm.RecordConnectionClosed()
+			},
+			expectedOpenedConnectionCount:  1,
+			expectedClosedConnectionCount:  1,
+			expectedAcceptConnectionErrors: 0,
+			expectedCloseConnectionErrors:  0,
+		},
+		{
+			description: "Count a close connection error",
+			testCase: func(pm *PrometheusMetrics) {
+				pm.RecordCloseConnectionErrors()
+			},
+			expectedOpenedConnectionCount:  1,
+			expectedClosedConnectionCount:  1,
+			expectedAcceptConnectionErrors: 0,
+			expectedCloseConnectionErrors:  1,
+		},
+		{
+			description: "Count an accept connection error",
+			testCase: func(pm *PrometheusMetrics) {
+				pm.RecordCloseConnectionErrors()
+				pm.RecordAcceptConnectionErrors()
+			},
+			expectedOpenedConnectionCount:  1,
+			expectedClosedConnectionCount:  1,
+			expectedAcceptConnectionErrors: 1,
+			expectedCloseConnectionErrors:  2,
+		},
+	}
+
+	m := createPrometheusMetricsForTesting()
+
+	for _, test := range testCases {
+		test.testCase(m)
+
+		assertCounterValue(t, test.description, m.Connections.ConnectionsOpened, test.expectedOpenedConnectionCount)
+		assertCounterValue(t, test.description, m.Connections.ConnectionsClosed, test.expectedClosedConnectionCount)
+
+		assertCounterVecValue(t, test.description, m.Connections.ConnectionsErrors, test.expectedAcceptConnectionErrors, prometheus.Labels{ConnErrorKey: AcceptVal})
+		assertCounterVecValue(t, test.description, m.Connections.ConnectionsErrors, test.expectedCloseConnectionErrors, prometheus.Labels{ConnErrorKey: CloseVal})
+	}
+}
+
+func TestExtraTTLMetrics(t *testing.T) {
+	m := createPrometheusMetricsForTesting()
+
+	assertHistogram(t, "Assert the extra time to live in seconds allocated when metrics were created", m.ExtraTTL.ExtraTTLSeconds, 1, 5000.00)
+
+	m.RecordExtraTTLSeconds(5)
+	assertHistogram(t, "Assert the extra time to live in seconds was logged", m.ExtraTTL.ExtraTTLSeconds, 2, 5005.00)
+}
+
+func TestMetricCountGatekeeping(t *testing.T) {
+	m := createPrometheusMetricsForTesting()
+
+	// Gather All Metrics
+	metricFamilies, err := m.Registry.Gather()
+	assert.NoError(t, err, "gather metics")
+
+	// Summarize By Adapter Cardinality
+	// - This requires metrics to be preloaded. We don't preload account metrics, so we can't test those.
+	generalCardinalityCount := 0
+
+	for _, metricFamily := range metricFamilies {
+		generalCardinalityCount += len(metricFamily.GetMetric())
+	}
+
+	// Verify General Cardinality
+	// - This assertion provides a warning for newly added high-cardinality non-adapter specific metrics. The hardcoded limit
+	//   is an arbitrary soft ceiling. Thought should be given as to the value of the new metrics if you find yourself
+	//   needing to increase this number.
+	assert.True(t, generalCardinalityCount <= 52, fmt.Sprintf("General Cardinality. Expected to hace less than 24 metrics, we have: %d \n", generalCardinalityCount))
```

Comment:
This test is now failing with the addition of the go metrics collectors. There's no need to make this an exact number. Its purpose is to protect against high cardinality. Consider setting to 100 or so as a rough limit.

Reply:
done